lib/api/accounts/user-accounts.js

```diff
@@ -0,0 +1,143 @@
+'use strict'
+
+const express = require('express')
+const bodyParser = require('body-parser').urlencoded({ extended: false })
+const debug = require('../../debug').accounts
+
+const CreateAccountRequest = require('../../requests/create-account-request')
+const AddCertificateRequest = require('../../requests/add-cert-request')
+
+/**
+ * Returns an Express middleware handler for checking if a particular account
+ * exists (used by Signup apps).
+ *
+ * @param accountManager {AccountManager}
+ *
+ * @return {Function}
+ */
+function checkAccountExists (accountManager) {
+  return (req, res, next) => {
+    let accountUri = req.hostname
+    accountManager.accountUriExists(accountUri)
+      .then(found => {
+        if (!found) {
+          debug(`Account ${accountUri} is available (for ${req.originalUrl})`)
+          return res.sendStatus(404)
+        }
+        debug(`Account ${accountUri} is not available (for ${req.originalUrl})`)
+        next()
+      })
+  }
+}
+
+/**
+ * Returns an Express middleware handler for creating a new user account
+ * (POST /api/accounts/new).
+ *
+ * @param accountManager {AccountManager}
+ *
+ * @return {Function}
+ */
+function createAccount (accountManager) {
+  return (req, res, next) => {
+    let request
+
+    try {
+      request = CreateAccountRequest.fromParams(req, res, accountManager)
+    } catch (err) {
+      err.status = err.status || 400
+      return next(err)
+    }
+
+    return request.createAccount()
+      .catch(err => {
+        err.status = err.status || 400
+        next(err)
+      })
```

Comment:
One thing that makes it opaque to me: the fact that the `CreateAccountRequest` object swallows up the request/response objects from express rather than returning back to the express handler and deferring to it to choose how to respond to the client.

The other thing contributing to the opaqueness here IMO is the several layers of classes that interact to get the job done. The account creation handler instantiates a class called `CreateAccountRequest` which has a `createAccount` method, and then calls are made to the `AccountManager`. This is somewhat a matter of taste, and anyways you've already built and tested this particular approach anyways.

That said, I find code easier to understand when it is built as a collection of functions which operate on a small set of data rather than many classes each with their own ad-hoc interface.